### ISSUE TITLE
502 Bad Gateway after Lua Auth

### ISSUE THREAD

**Issue by devgig** [zmartzone/lua-resty-openidc#353]
###### Environment

- lua-resty-openidc version (e.g. 1.7.2)
- OpenID Connect provider (e.g. Keycloak, Azure AD)
- openresty/openresty:1.15.8.3-2-bionic (Docker Image)

###### Expected behaviour
After authenticating and getting an ID_Token (which I am getting), I would expect to return to index.html.  This location / works without adding auth.lua.  As soon as I add it, I get the error.

```
 location / {

    access_by_lua_file lua/auth.lua;
     
   
    root /usr/local/openresty/nginx/html;
    index index.html index.htm; 
    try_files $uri $uri/ /index.html = 404;

  }
```

###### Actual behaviour
after getting this message OIDC Authorization Code Flow completed -> Redirecting to original URL (/)
it redirect back to /callback?code=f6d57c9c88aa46...
and I get a closed keepalive connection (104: Connection reset by peer)
```
2020/09/02 21:18:25 [debug] 6#6: *1 [lua] openidc.lua:1188: authenticate(): OIDC Authorization Code Flow completed -> Redirecting to original URL (/)
172.23.182.199 - - [02/Sep/2020:21:18:25 +0000] "GET /callback?code=f6d57c9c88aa46ed8cbb7f7107a85d1c&state=1f665e74318fe617f026141113d4ef03 HTTP/1.1" 302 151 "https://dais-sessionmanager-perf.ausvdc02.pcf.dell.com/sm" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.83 Safari/537.36"
2020/09/02 21:18:25 [info] 6#6: *1 client 172.23.182.199 closed keepalive connection (104: Connection reset by peer)
2020/09/02 21:18:25 [info] 6#6: *1 client 172.23.182.199 closed keepalive connection (104: Connection reset by peer)
```
###### Minimized example
*Minimal, complete configuration that reproduces the behavior.*

###### Configuration and NGINX server log files
*Config and logs for the minimized example, possibly provided as attachments.*

**Comment by devgig** [comment]
@bodewig 
I have confirmed the issue is in Pivotal Kubernetes Service (PKS) and it is using NSX-T or NGINX Ingress.  Basically any Ingress is causing this issue.  If I go at it with just an IP address all is good.  I am not for sure how to get around this K8s Ingress issue.

**Comment by bodewig** [comment]
Over in #316 you mention the session cookie was sharded into two cookies. It might be that your ingress is swallowing one of them.

The log snippet you are providing really doesn't tell us anything. What we see is your authorization code grant flow has finished by a call to `/callback`, lua-resty-openidc sends a redirect back to `/` (first line), this redirect is recorded inside the access log (second line) and then the client (likely your ingress) closes the keepalive HTTP connection.

You will need to look into the logs of the ingress controller and investigate what the client sees. One common reason for failing requests behind an nginx I've seen is the size of the HTTP headers. If your cookie is so big it needs to be sharded than the sum of all HTTP headers will likely exceed the default settings - [client_header_buffer_size](https://nginx.org/en/docs/http/ngx_http_core_module.html#client_header_buffer_size) for example. They may need to be increased for things to work. The logs of your ingress should tell you what is wrong, if there is anything wrong.

This really sounds like a problem you need to resolve together with the people working on your ingress as all we can do is guessing at this point.

**Comment by devgig** [comment]
@bodewig 
I agree with your assessment.  I tried another OIDC provider and everything is working.  I will find someone that can help with viewing Ingress Control logs or see if there is a way to push the logs to standard out so I can see what is going on.  What the buffer size should be seems to be a guessing game right now.

**Comment by dmilcevski** [comment]
Although the issue is closed, I want to give my two cents of how I solved the same issue I had, so it may help anybody that has the same issue. I had the same problem where cookie was split into two cookies because of its size and the total size was > `6KB`. The cluster admin that has control over our NGINX Ingress was bit slow in responding the increase the `client-Header_buffer_size`, so I looked for possible ways to overcome the solution and played around with annotation to the ingress I can pass as a user. After few tryouts I found out that the following configuration did the trick for me.

```
nginx.ingress.kubernetes.io/proxy-buffering: "on"
nginx.ingress.kubernetes.io/proxy-buffers-number: "8"
nginx.ingress.kubernetes.io/proxy-buffer-size: "16k"
```
